libsolidity/AST.h

```diff
@@ -315,6 +334,31 @@ class StructDefinition: public Declaration
 	std::vector<ASTPointer<VariableDeclaration>> m_members;
 };
 
+class EnumDefinition: public Declaration
+{
+public:
+	EnumDefinition(Location const& _location,
+				   ASTPointer<ASTString> const& _name,
+				   std::vector<ASTPointer<EnumValue>> const& _members):
+		Declaration(_location, _name), m_members(_members) {}
+	virtual void accept(ASTVisitor& _visitor) override;
+	virtual void accept(ASTConstVisitor& _visitor) const override;
+
+	std::vector<ASTPointer<EnumValue>> const& getMembers() const { return m_members; }
+
+	virtual TypePointer getType(ContractDefinition const*) const override;
+
+	/// @returns the value that the string has in the Enum
+	unsigned int getMemberValue(ASTString const& _member) const;
```

Comment:
Actually I am not happy about all the type checking being part of the AST - at some point I would like to move it to its own class. And also the type system has too much "implementation logic", it would be better if it only had abstract type information.